api_compatibility_policy.md

```diff
@@ -151,3 +151,68 @@ Tekton Pipelines may introduce breaking changes to its Go client libraries, as l
 do not impact users' yaml/json CRD definitions. For example, a change that renames a field of a CRD
 would need to follow the policy on [backwards incompatible changes](#backwards-incompatible-changes),
 but a change that renames the Go struct type for that field is allowed.
+
+## Notes for Developers
+
+The following are things developers should keep in mind for deprecations. These
+tips may not apply in every case, so use your best judgement! If you're not sure
+how these affect your deprecation or timeline, ask the maintainers.
+
+1. Assume default values are used.
+
+   Feature flags are an excellent way for users to opt-in and try out new
+   behavior before it goes live. However unless someone is trying to experiment
+   with the new feature, many users and other tools built on top of the API
+   won't know about it until it becomes the new default.
+
+   Try to give users as much time with a new feature flag in opt-out mode as you
+   can. Users may not update releases immediately, so this gives you more time
+   to get feedback from a broader set of users while still having an escape
+   hatch to revert back to old behavior if problems come up. This may mean
+   splitting up your deprecation window into opt-in and opt-out phases. For
+   particularly disruptive changes, consider making the switch to opt-out its
+   own deprecation notification.
+
+   A conservative deprecation sequence to move from field A -> B might look
+   like:
+
+   1. (inital state) Return field A
+   2. Introduce new field B with opt-in feature flag (disabled by default).
+   3. Set feature flag to return both A and B in responses by default.
+   4. Set feature flag to opt-in users by default and only return field B.
+   5. Remove feature flag.
+
+2. Use Godoc `Deprecated:` comments.
+
+   Go has a feature for using special comments to notify language servers / IDEs
+   that types and fields are deprecated. Use this as much as possible!
+
+   ```go
+   // Foo is a ...
+   //
+   // Deprecated: use Bar instead
+   type Foo struct {...}
+   ```
+
+   See https://github.com/golang/go/wiki/Deprecated for more details.
+
+3. Tombstone API fields before removal.
+
+   Removing fields from the API is most disruptive since it affects both the
+   server and clients, but **clients are not guaranteed to run at the same
+   version as the server**. This can break an integration's ability to safely
+   talk to older API servers.
+
+   A question to ask yourself is: "if I remove this field, can this client still
+   be used with the oldest supported LTS server?". If the answer is no, removing
```

Comment:
Thanks @wlynch 

Shall we suggest there to be a hard earliest timeline that the fields could be removed after defaulting the feature flag in one LTS? And I have been wondering if this also needs to be updated in the compatibility policy?

Reply:
I think this depends on how the field was deprecated.

Most conservative is wait until the last LTS where a user could downgrade cycles out (worst case 1 yr from release). At minimum I'd probably consider it a separate deprecation event from the feature flag removal.

There are probably things you could do to reduce this. For example for the embedded status deprecation - what made this more painful was that the field **might** not be there. If we guaranteed earlier that the childRefs would always be present (since it would have been an additive backwards compatible change) and scoped the deprecation around the opt-out flag we probably could have removed the feature flag and the API field at the same time since we could guarantee childrefs to always be present in older LTS versions.